File: output/CodeGen/avatar/Python/Java_handraft/atcoder_ABC127_F.java
public class atcoder_ABC127_F {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int q = in.nextInt();
        int[] p1 = new int[q];
        int[] p2 = new int[q];
        int[] b = new int[q];
        for (int i = 0; i < q; i++) {
            b[i] = in.nextInt();
        }
        for (int i = 0; i < q; i++) {
            p1[i] = in.nextInt();
        }
        for (int i = 0; i < q; i++) {
            p2[i] = in.nextInt();
        }
        int sum_b = 0;
        int sum_p1 = 0;
        int sum_p2 = 0;
        for (int i = 0; i < q; i++) {
            if (b[i] == 2) {
                if (p1.length == p2.length) {
                    System.out.println(-p2[0], sum_p1 - p1.length * -p2[0] + p2.length * -p2[0] - sum_p2 + sum_b);
                } else {
                    System.out.println(p1[0], sum_p1 - p1.length * p1[0] + p2.length * p1[0] - sum_p2 + sum_b);
                }
            } else {
                sum_b += b[i];
                if (p1.length == 0) {
                    heapq.heappush(p1, p1[i]);
                    sum_p1 += p1[i];
                } else if (p1[0] <= p1[i]) {
                    heapq.heappush(p1, p1[i]);
                    sum_p1 += p1[i];
                } else {
                    heapq.heappush(p2, -p1[i]);
                    sum_p2 += p1[i];
                }
                if (p1.length - 1 > p2.length) {
                    heapq.heappop(p2);
                    sum_p2 += p2[0];
                    sum_p1 -= p2[0];
                }
                if (p1.length - 1 > p2.length) {
                    heapq.heappop(p1);
                    sum_p1 -= p1[0];
                    sum_p2 += p1[0];
                }
            }
            if (p1.length - 1 > p2.length) {
                heapq.heappop(p2);
                sum_p2 += p2[0];
                sum_p1 -= p2[0];
            }
            if (p1.length - 1 > p2.length) {
                heapq.heappop(p1);
                sum_p1 -= p1[0];
                sum_p2 += p1[0];
            }
        }
    }
}

File: output/ChatGPT/avatar/Java/Python/atcoder_ABC127_F.py
import sys
import heapq

def main():
    t = int(sys.stdin.readline())
    pq_max = []
    pq_min = []
    sum_min = 0
    sum_max = 0
    sum_b = 0
    min_val = 0

    for _ in range(t):
        s = sys.stdin.readline().split()
        
        if s[0][0] == '2':
            ans = min_val * len(pq_min) - sum_min
            ans1 = min_val * len(pq_max) - sum_max
            sys.stdout.write(str(min_val) + " " + str(ans + ans1 + sum_b) + "\n")
        else:
            in_val = int(s[1])
            sum_b += int(s[2])
            
            if in_val > min_val:
                heapq.heappush(pq_max, in_val)
                sum_max += in_val
            else:
                heapq.heappush(pq_min, -in_val)
                sum_min -= in_val
            
            if len(pq_min) > len(pq_max):
                sum_max += pq_min[0]
                sum_min -= pq_min[0]
                heapq.heappush(pq_max, -heapq.heappop(pq_min))
            
            if len(pq_min) < len(pq_max):
                sum_max -= pq_max[0]
                sum_min += pq_max[0]
                heapq.heappush(pq_min, -heapq.heappop(pq_max))
            
            min_val = -pq_min[0]
    
    sys.stdout.flush()

if __name__ == "__main__":
    main()